client-common/src/main/java/com/cloudera/livy/client/common/ClientConf.java

```diff
@@ -193,4 +212,40 @@ private void check(boolean test, String message, Object... args) {
     }
   }
 
+  /** Logs a warning message if the given config key is deprecated. */
+  private void logDeprecationWarning(String key) {
+    DeprecatedConf altConfs = getConfigsWithAlternatives().get(key);
+    if (altConfs != null) {
+      LOG.warn("The configuration key " + altConfs.key() + " has been deprecated as of Livy "
+        + altConfs.version() + " and may be removed in the future. Please use the new key "
+        + key + " instead.");
+      return;
+    }
+
+    DeprecatedConf depConfs = getDeprecatedConfigs().get(key);
+    if (depConfs != null) {
+      LOG.warn("The configuration key " + depConfs.key() + " has been deprecated as of Livy "
+        + depConfs.version() + " and may be removed in the future. "
+        + depConfs.deprecationMessage());
+    }
+  }
+
+  /** Maps valid key to DeprecatedConf with the deprecated key. */
+  protected abstract Map<String, DeprecatedConf> getConfigsWithAlternatives();
+
+  /** Maps deprecated key to DeprecatedConf with the same key. */
+  protected abstract Map<String, DeprecatedConf> getDeprecatedConfigs();
```

Comment:
I think I see where you're going, but `DeprecatedConf.key()` is the deprecated key, the new key is the string that maps to `DeprecatedConf` in the `configsWithAlternatives` HashMap. This would mean changing how the new key maps to the deprecated key in order to allow deprecated keys with no alternative to exist in the same map since the key in the map would be either `null` or `""` and you couldn't have multiple deprecated confs without alternatives in the Map (by the def of a map).
I'm not saying having one Map isn't possible, but based on how I wrote the `DeprecatedConf` interface it would either require more complex code, or a reworking of the interface. I originally based this implementation on Spark's, which had two separate lists, but if you think having a single list would be better in the long run and @zjffdu or @alex-the-man agree I'm willing to rework it from the bottom up with that goal. I'm not super attached to this particular implementation, I just want to make sure future deprecation is straight forward and simple.